lightning/src/util/base32.rs

```diff
@@ -0,0 +1,368 @@
+// The original piece of software can be found at https://crates.io/crates/base32(v0.4.0)
+// The original portions of this software are Copyright (c) 2015 The base32 Developers
+
+// This file is licensed under either of
+// Apache License, Version 2.0, (LICENSE-APACHE or http://www.apache.org/licenses/LICENSE-2.0) or
+// MIT license (LICENSE-MIT or http://opensource.org/licenses/MIT) at your option.
+
+
+use crate::prelude::*;
+
+/// RFC4648 encoding table
+const RFC4648_ALPHABET: &'static [u8] = b"ABCDEFGHIJKLMNOPQRSTUVWXYZ234567";
+
+/// Zbase encoding alphabet
+const ZBASE_ALPHABET: &'static [u8] = b"ybndrfg8ejkmcpqxot1uwisza345h769";
+
+/// RFC4648 decoding table
+const RFC4648_INV_ALPHABET: [i8; 43] = [
+	-1, -1, 26, 27, 28, 29, 30, 31, -1, -1, -1, -1, -1, 0, -1, -1, -1, 0, 1, 2, 3, 4, 5, 6, 7, 8,
+	9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25,
+];
+
+/// Zbase decoding table
+const ZBASE_INV_ALPHABET: [i8; 43] = [
+	-1, 18, -1, 25, 26, 27, 30, 29, 7, 31, -1, -1, -1, -1, -1, -1, -1, 24, 1, 12, 3, 8, 5, 6, 28,
+	21, 9, 10, -1, 11, 2, 16, 13, 14, 4, 22, 17, 19, -1, 20, 15, 0, 23,
+];
+
+/// Alphabet used for encoding and decoding.
+#[derive(Copy, Clone)]
+pub enum Alphabet {
+	/// RFC4648 encoding.
+	RFC4648 {
+		/// Whether to use padding.
+		padding: bool
+	},
+	/// Zbase32 encoding.
+	ZBase32
+}
+
+impl Alphabet {
+	/// Encode bytes into a base32 string.
+	pub fn encode(&self, data: &[u8]) -> Result<String, ()> {
+		let ret = match self {
+			Self::RFC4648 { padding } =>{
+				Self::rfc4648_encode(data, *padding)
+			},
+			Self::ZBase32 => {
+				Self::zbase32_encode(data)
+			}
+		};
+		match String::from_utf8(ret) {
+			Ok(s) => Ok(s),
+			Err(_) => Err(())
+		}
+	}
+
+	/// Decode a base32 string into a byte vector.
+	pub fn decode(&self, data: &str) -> Result<Vec<u8>, ()> {
+		match self {
+			Self::RFC4648 { .. } => {
+				Self::rfc4648_decode(data.to_ascii_uppercase().as_bytes())
+			},
+			Self::ZBase32 => {
+				Self::zbase32_decode(data.as_bytes())
+			},
+		}
+	}
+
+	fn rfc4648_encode(data: &[u8], padding: bool) -> Vec<u8> {
+		let cap = (data.len() + 3) / 4 * 5;
+		let mut ret = Self::encode_data(cap, data, RFC4648_ALPHABET);
+		if data.len() % 5 != 0 {
+			let len = ret.len();
+			let num_extra = 8 - (data.len() % 5 * 8 + 4) / 5;
+			if padding {
+				for i in 1..num_extra + 1 {
+					ret[len - i] = b'=';
+				}
+			} else {
+				ret.truncate(len - num_extra);
+			}
+		}
+		ret
+	}
+
+	fn zbase32_encode(data: &[u8])-> Vec<u8> {
+		let cap = (data.len() + 4) / 5 * 8;
+		let mut ret = Self::encode_data(cap, data, ZBASE_ALPHABET);
+		ret.truncate((data.len() * 8 + 4) / 5);
+		// Check that our capacity calculation doesn't under-shoot in fuzzing
+		#[cfg(fuzzing)]
+		assert_eq!(ret.capacity(), cap);
+		ret
+	}
+
+	fn validate_alphabet_char(&c: &u8, alphabet: [i8; 43], padding: &mut usize) -> Result<(), ()> {
+		if alphabet == RFC4648_INV_ALPHABET {
```

Comment:
Hmm tbh the main reason I didnt add the zbase validation here is because its working and I didnt wanna change/move it.
Regarding the validation itself, they use different alphabets and the rfc4648 is using padding so I think we need to differentiate here. 

Looking into the code, I agree that its weird that we validate only one case in the "validation" function.

Reply:
Another way to approach it is to change `decode_data` signature to accept `Option<impl Fn(&alphaet) -> bool`, and then each different alphabet can decide if to pass a validator down to it.

Reply:
@tnull updated the alphabet_char_validator to include zbase as well